[#200] Respond HTTP 204 when no more changes are available

iperdomo opened the pull request:
The proxy server (nginx) will cache this response for 1h. Useful for _eager_ clients re-requesting the same thing over and over again. The value of 1h is arbitrary and we can adjust it.

A client can bypass the proxy by using a request header `Pragma: no-cache`

iperdomo committed:
[#200] Respond HTTP 204 when no more changes are available

The proxy server (nginx) will cache this response for 1h. Useful for
_eager_ clients re-requesting the same thing over and over again.
The value of 1h is arbitrary and we can adjust it.

A client can bypass the proxy by using a request header `Pragma: no-cache`

dlebrero commented on nginx-auth0/nginx.conf:
```diff
@@ -155,12 +160,24 @@ http {
+      proxy_cache http_204;
+      proxy_cache_revalidate off;
+      proxy_cache_background_update on;
+      proxy_cache_lock on;
```
Love it! It makes me wonder if we should look at this (or similar) for the whole API (as another Github issue).

Can we set also `proxy_cache_lock_timeout` and `proxy_cache_lock_age`? I think the defaults are too low for the latency that I expect the sync API will have if there is any data to be fetched.

iperdomo replied:
> It makes me wonder if we should look at this (or similar) for the whole API (as another Github issue). - https://github.com/akvo/akvo-flow-api/issues/204

iperdomo replied:
> Can we set also `proxy_cache_lock_timeout` and `proxy_cache_lock_age`

Yes

dlebrero commented on nginx-auth0/nginx.conf:
```diff
@@ -155,12 +160,24 @@ http {
+      proxy_cache_lock on;
+      proxy_cache_methods GET;
+      proxy_cache_key $proxy_host$request_uri$akvoemail;
+      proxy_cache_valid 204 60m;
```
I was thinking more in 30 to 60 secs. "Real time"

iperdomo replied:
For _"real time"_ we have a way to _punch a whole_ and bypass the cache. My intention with the 1h cache is to avoid reaching the database and other services in a transparent way. The client is not aware that we're serving cached content.

dlebrero commented on nginx-auth0/nginx.conf:
```diff
@@ -155,12 +160,24 @@ http {
+      proxy_cache_background_update on;
+      proxy_cache_lock on;
+      proxy_cache_methods GET;
+      proxy_cache_key $proxy_host$request_uri$akvoemail;
```
Using `$request_uri` means that the user can use any cache busting param to bypass the cache. 

Is it possible to pick up just the http params that we support? This will cause some coupling between the backend code and the nginx one, so maybe I am just being paranoid and it is not really needed.

iperdomo replied:
I don't think we need to invest way too much on hardening the key based on _supported parameters_. We wanted something easier to implement than _Long poll_.

My intention is not to document that there is caching going on. The contract will be that after getting HTTP 204, the client must stop requesting and _wait_ on its side. This caching will be just a safety net for those clients that will not wait on their side.

dlebrero commented on nginx-auth0/nginx.conf:
```diff
@@ -155,12 +160,24 @@ http {
+      proxy_cache_methods GET;
+      proxy_cache_key $proxy_host$request_uri$akvoemail;
+      proxy_cache_valid 204 60m;
+      proxy_cache_bypass $http_pragma;
```
Same as the comment regarding `$request_uri`. Allow the client to "cheat". Is it going to be abused?

Assuming that the cache value is 30-60 secs, what is the benefit of using it?

iperdomo replied:
We're not going to document that you can bypass and force to hit the upstream server

iperdomo replied:
> Assuming that the cache value is 30-60 secs, what is the benefit of using it?

As mentioned in the previous comment, my intention is to have a larger cache time than 30-60s (for those bad behaving clients) with a _escape hatch_ for *our* services.

dlebrero commented on nginx-auth0/nginx.conf:
```diff
@@ -34,6 +34,9 @@ http {
   lua_capture_error_log 32m;
 
+  # Cache 204 responses
+  proxy_cache_path /data/nginx/cache levels=1:2 keys_zone=http_204:5m max_size=1g inactive=60m use_temp_path=off;
```
What is the plan for the non-auth0 version of the proxy?

iperdomo replied:
Good question. Need to think about it.

dlebrero commented on api/src/clojure/org/akvo/flow_api/endpoint/sync.clj:
```diff
@@ -41,20 +41,24 @@
-                           :next-sync-url (next-sync-url (utils/get-api-root req) alias (:unilog-id changes))}))
+              (if (= offset (unilog/get-cursor db-spec)) ;; end of the log
+                (-> (status {} 204)
+                    (header "Cache-Control" "max-age=3600"))
```
Isn't the Nginx adding this header? (Haven't tested it)

In any case, there is some duplication here and in the nginx code regarding the caching time. Either we make the `proxy_cache_valid` follow this header or we remove this header.

iperdomo replied:
According to the documentation, nginx will honor the `Cache-Control` header from the upstream server. I'll double check what is the behavior when both properties differ.

iperdomo replied:
> Isn't the Nginx adding this header? (Haven't tested it)

nginx is not adding the header, but it does cache the HTTP 204 response. I can see the instrumentation header:
````
< HTTP/1.1 204 No Content
< Server: openresty
< Date: Tue, 28 Jan 2020 09:24:49 GMT
< Connection: keep-alive
< Access-Control-Allow-Origin: *
< Access-Control-Allow-Credentials: true
< Access-Control-Allow-Methods: GET, OPTIONS
< Access-Control-Allow-Headers: Accept,Accept-Encoding,Accept-Language,Authorization,Connection,DNT,Host,Origin,Referer,User-Agent
< X-Cache-Status: MISS
< 

[...]

< HTTP/1.1 204 No Content
< Server: openresty
< Date: Tue, 28 Jan 2020 09:25:06 GMT
< Connection: keep-alive
< Access-Control-Allow-Origin: *
< Access-Control-Allow-Credentials: true
< Access-Control-Allow-Methods: GET, OPTIONS
< Access-Control-Allow-Headers: Accept,Accept-Encoding,Accept-Language,Authorization,Connection,DNT,Host,Origin,Referer,User-Agent
< X-Cache-Status: HIT

````
In this case the server is not sending the `Cache-Control` but we do have the configuration ` proxy_cache_valid 204 <time>`

dlebrero commented on api/src/clojure/org/akvo/flow_api/endpoint/sync.clj:
```diff
@@ -41,20 +41,24 @@
-                             (update :data-point-deleted #(map str %)))]
-                (response {:changes (dissoc changes :unilog-id)
-                           :next-sync-url (next-sync-url (utils/get-api-root req) alias (:unilog-id changes))}))
+              (if (= offset (unilog/get-cursor db-spec)) ;; end of the log
```
This is the third JDBC query: 1 for valid, 1 for last cursor and 1 for reading the data. 

Not an issue if we open the DB connection once, but if we open/close for each, maybe we should look into avoiding 2 of those 3 calls.

Being very picky here, so feel free to ignore.

iperdomo replied:
* the last cursor is one code path (`initial=true`)
* the valid and reading (`next=true`), for checking valid and reading data can be transformed into one

dlebrero commented on api/src/clojure/org/akvo/flow_api/endpoint/sync.clj:
```diff
@@ -41,20 +41,24 @@
                 db-spec (-> deps :unilog-db :spec (assoc :db-name db-name))
```
This function is becoming a monster. Maybe time to clean up?

iperdomo replied:
agreed :+1:

dlebrero reviewed (changes requested):
Good. I like the approach

iperdomo committed:
[#200] Reorder and clean up CI build process

iperdomo committed:
[#200] Disallow /sync with Keycloak based authentication

iperdomo committed:
[#200] Use Cache-Control header from upstream to control caching

* Test caching as part of the CI build

iperdomo committed:
[#200] Use a different network name for cache testing

iperdomo committed:
[#200] Decrease cache ttl to 60s

iperdomo committed:
[#200] Use a a single open connection across request

iperdomo committed:
[#200] Add test for X-Cache-Status BYPASS

iperdomo committed:
[#200] Split changes function

iperdomo committed:
[#200] Increase the values for cache_lock_timeout|cache_lock_age

* It's quite unlikely that the same user (part of the cache key)
  is requesting the same URL at the same time

dlebrero commented on api/src/clojure/org/akvo/flow_api/unilog/unilog.clj:
```diff
@@ -147,3 +147,14 @@
+  (fn [request]
+    (let [db-name (get-db-name (:instance-id request))
+          db-spec (-> unilog-db :spec (assoc :db-name db-name) event-log-spec)]
+      (with-open [connection (jdbc/get-connection  db-spec)]
```
Suggestion: use the with-db-connection macro. The docs discourage using get-connection, plus I guess that using the macro means that we do not need to manually build the {:connection} map

dlebrero commented on ci/build.sh:
```diff
@@ -23,40 +23,53 @@ fi
+    ./test.sh "http://localhost:8082/flow/" 2>&1 | grep 'X-Cache-Status: MISS'
+    ./test.sh "http://localhost:8082/flow/" 2>&1 | grep 'X-Cache-Status: HIT'
+    ./test.sh "http://localhost:8082/flow/" 2>&1 | grep 'X-Cache-Status: HIT'
+    ./test.sh "http://localhost:8082/flow/" --header "Pragma: no-cache" 2>&1 | grep 'X-Cache-Status: BYPASS'
```
Suggestion: I thought we agreed to remove the bypass as we have reduced the cache time to 60 secs, plus we can bypass it by using a random request param

dlebrero commented on ci/build.sh:
```diff
@@ -23,40 +23,53 @@ fi
+       "akvo/flow-api-auth0-proxy" -t -c /usr/local/openresty/nginx/conf/nginx.conf
 
+# Checking nginx caching
+(
```
Suggestion: Move out of build

dlebrero commented on ci/build.sh:
```diff
@@ -23,40 +23,53 @@ fi
+(
+    cd nginx-auth0
+    docker-compose up -d
+    sleep 5
```
Change: No sleeps in tests please.

dlebrero replied:
did we just deleted the sleep? Why is no longer needed?

iperdomo replied:
it seems that you're checking an _outdated_ code. This checks are *not* part of the build.sh anymore as you suggested.

> Suggestion: Move out of build

dlebrero commented on api/src/clojure/org/akvo/flow_api/endpoint/sync.clj:
```diff
@@ -26,35 +26,51 @@
+
 (defn changes [deps {:keys [alias instance-id params] :as req}]
   (let [{:keys [initial cursor next]} (spec/validate-params params-spec params)]
     (if (and initial (or cursor next))
```
Suggestion: change nested ifs for cond

dlebrero commented on api/src/clojure/org/akvo/flow_api/endpoint/sync.clj:
```diff
@@ -26,35 +26,51 @@
 
+(def no-more-changes (-> (status {} 204) (header "Cache-Control" "max-age=60")))
+
+
```
Suggestion: Remove empty line

dlebrero commented on api/src/clojure/org/akvo/flow_api/unilog/unilog.clj:
```diff
@@ -147,3 +147,14 @@
           data-point-changed (doall (data-point/by-ids ds (:data-point-changed events-2)))]
       (assoc events-2 :form-instance-changed form-instances :data-point-changed data-point-changed))))
+
+
```
Suggestion: remove empty line 🤐

dlebrero reviewed (approved):
Looks good. I would just ask for removing the “sleep” if it is not a lot of work as it always ends up making the test flaky and slower.

Added a bunch of optional suggestions. Feel free to ignore, no need to justify why.

Thanks!

iperdomo committed:
[#200] Remove extra empty lines

iperdomo committed:
[#200] Move test cache to its own script

iperdomo committed:
[#200] Remove `Pragma: no-cache` capability

iperdomo committed:
[#200] Use with-db-connection macro

Instead of creating the connection and then creating a `{:connection conn}`
map when needed, we can use the with-db-connection macro

iperdomo committed:
[#200] Refactor nested `(if)` to `(cond)`

iperdomo commented:
@dlebrero gave his :+1: via another channel

dlebrero commented:
I understand now the miscommunication.

By “move out” I actually meant “move this code out of the build.sh into another script and call this new script from the build.sh”

I did not realized in the code review that we were not doing calling the new script.

In general, I would never ever ever suggest removing any automated tests. If I do, I will give a very long explanation of why.

Apologies for not being more specific in the first place.


Sent from my iPhone

> On 30 Jan 2020, at 11:27, Iván Perdomo <notifications@github.com> wrote:
> 
> ﻿
> @iperdomo commented on this pull request.
> 
> In ci/build.sh:
> 
> >  
> +# Checking nginx caching
> +(
> +    cd nginx-auth0
> +    docker-compose up -d
> +    sleep 5
> it seems that you're checking an outdated code. This checks are not part of the build.sh anymore as you suggested.
> 
> Suggestion: Move out of build
> 
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub, or unsubscribe.
